### PR TITLE
Handle metrics

### PR DESCRIPTION
**Pull request by rmarianski**
Connects to https://github.com/mapzen/tile-tasks/issues/173

@zerebubuth, @iandees could you review please?

### CODE REVIEW

File: tapalcatl_server/handler.go [hunk 1]
```diff
@@ -25,23 +28,24 @@ type BufferManager interface {
...
+// the reqState structures and string generation serve to emit a single log entry line
 // a log parser will pick this up and use it to persist metrics
 // the string functions here are specific to the format used and should be updated with care
 
-type logResponseState int32
+type reqResponseState int32
 
 const (
-	ResponseState_Nil logResponseState = iota
+	ResponseState_Nil reqResponseState = iota
```
Comment (zerebubuth) [inline tapalcatl_server/handler.go]: Do we need the "nil" response state? As far as I can tell, it's not used anywhere in the code?

File: tapalcatl_server/handler.go [hunk 2]
```diff
@@ -149,19 +154,126 @@ func (logState *logEntryState) String() string {
...
+	lastMod metrics.Counter
+	etag    metrics.Counter
+
+	zipErr       metrics.Counter
+	respWriteErr metrics.Counter
+	condErr      metrics.Counter
+}
+
+func (smw *statsdMetricsWriter) Write(reqState *requestState) {
+	respStateInt := int(reqState.responseState)
```
Comment (zerebubuth) [inline tapalcatl_server/handler.go]: Nitpick: We've aliased `reqResponseState` to `int32`, so we really ought to use the same type here.

File: tapalcatl_server/handler.go (hunk 1 above)
Comment (rmarianski) [inline tapalcatl_server/handler.go]: > Do we need the "nil" response state? As far as I can tell, it's not used anywhere in the code?

It was meant to serve as the "zero state", or signal that the value hasn't been set. Would renaming it to something more explicit like `Unset` be better? Is this notion in general bad golang form?

File: tapalcatl_server/handler.go (hunk 2 above)
Comment (rmarianski) [inline tapalcatl_server/handler.go]: Good catch. Updated in 5b1fc10.

File: tapalcatl_server/handler.go (hunk 1 above)
Comment (zerebubuth) [inline tapalcatl_server/handler.go]: I'm not sure. I think it is considered good practice to have a "zero state" to be the default initialised value, but I'm not sure how that works when that value isn't meaningful. What does it mean if we get to the end of `MetatileHandler`'s `HandlerFunc` and we haven't set the value to something other than `ResponseState_Nil`?

Ideally, we could have the compiler check that no such code path exists. But I can't think of a way to do that, so perhaps we should log `ResponseState_Nil` in the `defer` block as an error?

Comment (rmarianski) [inline tapalcatl_server/handler.go]: > I'm not sure. I think it is considered good practice to have a "zero state" to be the default initialised value, but I'm not sure how that works when that value isn't meaningful. What does it mean if we get to the end of MetatileHandler's HandlerFunc and we haven't set the value to something other than ResponseState_Nil?

Yea, that's why I made the meaninglessness explicit :) It seemed better than having the first value default to either error or success as a code error and not be able to tell downstream.

> Ideally, we could have the compiler check that no such code path exists. But I can't think of a way to do that, so perhaps we should log ResponseState_Nil in the defer block as an error?

Another approach is to only allow the creation of these objects through a function, which takes the state explicitly. But for the compiler to enforce that we would need to shuffle packages around. I actually considered that anyway just for organizational reasons, but I think it's nice for the request processing state to sit close to the handler.

I logged an error in 0ba6a5e.

Comment (zerebubuth) [inline tapalcatl_server/handler.go]: > I logged an error in 0ba6a5e.

Cool, thanks! That's probably the best thing.

**Comment by rmarianski** [comment]
I'll go ahead and merge this as is now. There's been more discussion about the implementation in the issue since this pr was created, and if we decide to change our minds we can make a new pr to move in that direction.

**Pull request merged**